include/realtime_tools/realtime_box.h

```diff
@@ -27,51 +28,271 @@
 // POSSIBILITY OF SUCH DAMAGE.
 
 // Author: Stuart Glaser
+// Author: Lennart Nachtigall
 
-#ifndef REALTIME_TOOLS__REALTIME_BOX_H__
-#define REALTIME_TOOLS__REALTIME_BOX_H__
+#ifndef REALTIME_TOOLS__REALTIME_BOX_H_
+#define REALTIME_TOOLS__REALTIME_BOX_H_
 
+#include <functional>
+#include <initializer_list>
 #include <mutex>
-#include <string>
+#include <optional>
+#include <utility>
+
+#include <rcpputils/pointer_traits.hpp>
 
 namespace realtime_tools
 {
-/*!
 
-  Strongly suggested that you use an std::shared_ptr in this box to
-  guarantee realtime safety.
+template <typename T>
+constexpr auto is_ptr_or_smart_ptr = rcpputils::is_pointer<T>::value;
+
+/*!
+    A Box that ensures thread safe access to the boxed contents.
+    Access is best effort. If it can not lock it will return.
 
- */
-template <class T>
-class RealtimeBox
+    NOTE about pointers:
+    You can use pointers with this box but the access will be different.
+    Only use the get/set methods that take function pointer for accessing the internal value.
+*/
+template <class T, typename mutex_type = std::mutex>
+class RealtimeBoxBase
 {
+  static_assert(std::is_copy_constructible_v<T>, "Passed type must be copy constructible");
+
 public:
-  explicit RealtimeBox(const T & initial = T()) : thing_(initial) {}
+  using mutex_t = mutex_type;
+  using type = T;
+  // Provide various constructors
+  constexpr explicit RealtimeBoxBase(const T & init = T{}) : value_(init) {}
+  constexpr explicit RealtimeBoxBase(const T && init) : value_(std::move(init)) {}
 
-  void set(const T & value)
+  // Copy constructor
+  constexpr RealtimeBoxBase(const RealtimeBoxBase & o)
   {
-    std::lock_guard<std::mutex> guard(thing_lock_RT_);
-    thing_ = value;
+    // Lock the other box mutex
+    std::unique_lock<mutex_t> lock(o.lock_);
+    // We do not need to lock our own mutex because we are currently in the process of being created
+    value_ = o.value_;
   }
+  // Copy assignment constructor
+  constexpr RealtimeBoxBase & operator=(const RealtimeBoxBase & o)
+  {
+    // Check for self assignment (and a potential deadlock)
+    if (&o != this) {
+      // Lock the other box mutex
+      std::unique_lock<mutex_t> lock_other(o.lock_);
+      std::unique_lock<mutex_t> lock_self(lock_);
 
-  void get(T & ref)
+      value_ = o.value_;
+    }
+    return *this;
+  }
+  constexpr RealtimeBoxBase(RealtimeBoxBase && o)
   {
-    std::lock_guard<std::mutex> guard(thing_lock_RT_);
-    ref = thing_;
+    // Lock the other box mutex
+    std::unique_lock<mutex_t> lock(o.lock_);
+    // We do not need to lock our own mutex because we are currently in the process of being created
+    value_ = std::move(o.value_);
   }
+  // Only enabled for types that can be constructed from an initializer list
+  template <typename U = T>
+  constexpr RealtimeBoxBase(
+    const std::initializer_list<U> & init,
+    std::enable_if_t<std::is_constructible_v<U, std::initializer_list>>)
```

Comment:
```suggestion
    std::enable_if_t<std::is_constructible_v<U, std::initializer_list<U>>>)
```
is this correct, see #183?